spacy/ml/callbacks.py

```diff
@@ -11,29 +16,109 @@
     from ..language import Language  # noqa: F401
 
 
+DEFAULT_ANNOTATABLE_PIPE_METHODS = [
+    "pipe",
+    "predict",
+    "set_annotations",
+    "update",
+    "rehearse",
+    "get_loss",
+    "initialize",
+    "begin_update",
+    "finish_update",
+    "update",
+]
+
+
+def models_with_nvtx_range(nlp, forward_color: int, backprop_color: int):
+    pipes = [
+        pipe
+        for _, pipe in nlp.components
+        if hasattr(pipe, "is_trainable") and pipe.is_trainable
+    ]
+
+    # We need process all models jointly to avoid wrapping callbacks twice.
+    models: Model = Model(
+        "wrap_with_nvtx_range",
+        forward=lambda model, X, is_train: ...,
+        layers=[pipe.model for pipe in pipes],
+    )
```

Comment:
This feels slightly hacky to me, to make a `Model` just to be able to use the `walk` method. Can't we keep a `Set[int]` of seen IDs across models just like the `walk` method would do, but without the additional overhead of creating an actual artificial `Model`?

Reply:
This code wasn't originally introduced in this PR - I just had to move it out of the original callback function to accommodate the new ensemble callback.

Seems like the usage of `Model` is intended, according to [this commit](https://github.com/explosion/spaCy/pull/9124/commits/071d930bf222a9768a903ff0bbacb5411c68222d).

Reply:
Before we had 
```
models = wrap_model_recursive(models, with_nvtx_range)
```
with `wrap_model_recursive` calling 
```
    for node in list(models.walk()):
        models.replace_node(node, wrapper(node))
```
Now, we call
```
    for node in models.walk():
        with_nvtx_range(
            node, forward_color=forward_color, backprop_color=backprop_color
        )
```
directly, using the artificial `models` of type `Model` only to use the `walk` method, which is nothing more than
```
        queue = [self]
        seen: Set[int] = set()
        for node in queue:
            if id(node) in seen:
                continue
            seen.add(id(node))
            yield node
            queue.extend(node.layers)
```
So basically what I'm saying is, can't we do this more directly, without the artificial `Model`, walk over all the nodes in the different `pipe.model for pipe in pipes`, keep a set of `seen` IDs and avoid calling the  `with_nvtx_range` method if that layer has already been seen?

I'm proposing this only because there's a `Model` constructor used here but we never really a `Model` - we just want a convenient way to iterate over nodes without duplication, right?

Reply:
Ah, gotcha! Dunno why I got confused about that :sweat_smile: Yeah, you're right - we can get rid of that wrapper instance of `Model`.